File: likho-app/src/lib/backup.ts
/**
 * Backup Service
 * 
 * Provides comprehensive backup and restore functionality for the Likho app.
 * Supports SQLite, JSON, and Gzip formats with automatic scheduling.
 */

import { invoke } from "@tauri-apps/api/core";
import { save, open } from "@tauri-apps/plugin-dialog";
import { homeDir, join } from "@tauri-apps/api/path";

// Types matching Rust backend
export type BackupFormat = "sqlite" | "json" | "gzip";

export interface BackupMetadata {
  version: string;
  created_at: string;
  original_size: number;
  compressed_size: number;
  format: string;
  checksum: string;
  include_embeddings: boolean;
  app_version: string;
}

export interface BackupInfo {
  path: string;
  filename: string;
  size: number;
  created_at: string;
  metadata?: BackupMetadata;
}

export interface RestorePreview {
  backup_date: string;
  original_size: number;
  format: string;
  version: string;
  warnings: string[];
}

export interface AutoBackupSettings {
  enabled: boolean;
  frequency: "daily" | "weekly" | "monthly";
  backup_dir: string;
  max_backups: number;
  include_embeddings: boolean;
  format: BackupFormat;
}

export interface AutoBackupStatus {
  enabled: boolean;
  last_backup: string | null;
  next_backup: string | null;
  backups_count: number;
  total_size: number;
}

export interface BackupProgress {
  stage: "preparing" | "exporting" | "compressing" | "finalizing" | "completed" | "error";
  progress: number;
  message: string;
  error?: string;
}

// Default settings
const DEFAULT_SETTINGS: AutoBackupSettings = {
  enabled: false,
  frequency: "weekly",
  backup_dir: "",
  max_backups: 10,
  include_embeddings: true,
  format: "sqlite",
};

// Settings storage key
const SETTINGS_KEY = "likho_auto_backup_settings";

/**
 * Get the default backup directory
 */
export async function getDefaultBackupDir(): Promise<string> {
  const home = await homeDir();
  return await join(home, "Likho", "Backups");
}

/**
 * Load auto-backup settings from localStorage
 */
export function loadBackupSettings(): AutoBackupSettings {
  try {
    const stored = localStorage.getItem(SETTINGS_KEY);
    if (stored) {
      return { ...DEFAULT_SETTINGS, ...JSON.parse(stored) };
    }
  } catch {
    console.error("Failed to load backup settings");
  }
  return { ...DEFAULT_SETTINGS };
}

/**
 * Save auto-backup settings to localStorage
 */
export function saveBackupSettings(settings: AutoBackupSettings): void {
  try {
    localStorage.setItem(SETTINGS_KEY, JSON.stringify(settings));
  } catch {
    console.error("Failed to save backup settings");
  }
}

/**
 * Create a manual backup
 * Opens a save dialog and exports the database
 */
export async function createBackup(
  format: BackupFormat = "sqlite",
  includeEmbeddings: boolean = true,
  onProgress?: (progress: BackupProgress) => void
): Promise<{ success: boolean; path?: string; metadata?: BackupMetadata; error?: string }> {
  try {
    onProgress?.({
      stage: "preparing",
      progress: 0,
      message: "Preparing backup...",
    });

    // Get default file name
    const timestamp = new Date().toISOString().replace(/[:.]/g, "-").slice(0, 19);
    const extension = format === "json" ? "json" : format === "gzip" ? "db.gz" : "db";
    const defaultName = `likho_backup_${timestamp}.${extension}`;

    onProgress?.({
      stage: "preparing",
      progress: 20,
      message: "Opening save dialog...",
    });

    // Open save dialog
    const savePath = await save({
      defaultPath: defaultName,
      filters: [
        { name: "SQLite Database", extensions: ["db"] },
        { name: "JSON Backup", extensions: ["json"] },
        { name: "Gzip Compressed", extensions: ["gz", "db.gz"] },
        { name: "All Files", extensions: ["*"] },
      ],
      title: "Save Likho Backup",
    });

    if (!savePath) {
      return { success: false, error: "Save cancelled by user" };
    }

    onProgress?.({
      stage: "exporting",
      progress: 30,
      message: "Exporting database...",
    });

    // Call Rust backend to export
    const metadata = await invoke<BackupMetadata>("export_database", {
      request: {
        targetPath: savePath,
        format,
        includeEmbeddings,
      },
    });

    onProgress?.({
      stage: "completed",
      progress: 100,
      message: "Backup completed successfully!",
    });

    return { success: true, path: savePath, metadata };
  } catch (error) {
    const errorMessage = error instanceof Error ? error.message : String(error);
    onProgress?.({
      stage: "error",
      progress: 0,
      message: "Backup failed",
      error: errorMessage,
    });
    return { success: false, error: errorMessage };
  }
}

/**
 * Preview a restore operation
 * Returns information about the backup without performing the restore
 */
export async function previewRestore(
  filePath: string
): Promise<{ success: boolean; preview?: RestorePreview; error?: string }> {
  try {
    const preview = await invoke<RestorePreview>("preview_restore", {
      request: { sourcePath: filePath },
    });
    return { success: true, preview };
  } catch (error) {
    const errorMessage = error instanceof Error ? error.message : String(error);
    return { success: false, error: errorMessage };
  }
}

/**
 * Restore database from a backup file
 * WARNING: This will overwrite current data!
 */
export async function restoreBackup(
  filePath: string,
  onProgress?: (progress: BackupProgress) => void
): Promise<{ success: boolean; error?: string }> {
  try {
    onProgress?.({
      stage: "preparing",
      progress: 0,
      message: "Preparing restore...",
    });

    // First verify the backup
    onProgress?.({
      stage: "preparing",
      progress: 10,
      message: "Verifying backup integrity...",
    });

    const isValid = await invoke<boolean>("verify_backup_integrity", {
      path: filePath,
    });

    if (!isValid) {
      throw new Error("Backup file is corrupted or invalid");
    }

    onProgress?.({
      stage: "exporting",
      progress: 30,
      message: "Restoring database...",
    });

    // Execute restore
    await invoke("execute_restore", {
      request: { sourcePath: filePath },
    });

    onProgress?.({
      stage: "completed",
      progress: 100,
      message: "Restore completed successfully!",
    });

    return { success: true };
  } catch (error) {
    const errorMessage = error instanceof Error ? error.message : String(error);
    onProgress?.({
      stage: "error",
      progress: 0,
      message: "Restore failed",
      error: errorMessage,
    });
    return { success: false, error: errorMessage };
  }
}

/**
 * Open file dialog to select a backup file for restore
 */
export async function selectBackupFile(): Promise<string | null> {
  const selected = await open({
    multiple: false,
    filters: [
      { name: "Backup Files", extensions: ["db", "json", "gz", "db.gz"] },
      { name: "SQLite Database", extensions: ["db"] },
      { name: "JSON Backup", extensions: ["json"] },
      { name: "Gzip Compressed", extensions: ["gz", "db.gz"] },
      { name: "All Files", extensions: ["*"] },
    ],
    title: "Select Likho Backup",
  });

  return selected as string | null;
}

/**
 * Verify backup file integrity
 */
export async function verifyBackupIntegrity(filePath: string): Promise<boolean> {
  try {
    return await invoke<boolean>("verify_backup_integrity", { path: filePath });
  } catch {
    return false;
  }
}

/**
 * List all backups in a directory
 */
export async function listBackups(backupDir: string): Promise<BackupInfo[]> {
  try {
    return await invoke<BackupInfo[]>("list_backups", { backupDir });
  } catch (error) {
    console.error("Failed to list backups:", error);
    return [];
  }
}

/**
 * Delete a backup file
 */
export async function deleteBackup(filePath: string): Promise<{ success: boolean; error?: string }> {
  try {
    await invoke("delete_backup", { path: filePath });
    return { success: true };
  } catch (error) {
    const errorMessage = error instanceof Error ? error.message : String(error);
    return { success: false, error: errorMessage };
  }
}

/**
 * Get auto-backup status
 */
export async function getAutoBackupStatus(): Promise<AutoBackupStatus> {
  try {
    return await invoke<AutoBackupStatus>("get_auto_backup_status");
  } catch (error) {
    console.error("Failed to get auto-backup status:", error);
    return {
      enabled: false,
      last_backup: null,
      next_backup: null,
      backups_count: 0,
      total_size: 0,
    };
  }
}

/**
 * Update auto-backup settings
 */
export async function setAutoBackupSettings(
  settings: AutoBackupSettings
): Promise<{ success: boolean; error?: string }> {
  try {
    await invoke("set_auto_backup_settings", { settings });
    saveBackupSettings(settings);
    return { success: true };
  } catch (error) {
    const errorMessage = error instanceof Error ? error.message : String(error);
    return { success: false, error: errorMessage };
  }
}

/**
 * Trigger an auto-backup if needed (based on schedule)
 * Returns the metadata if a backup was created, null if not needed
 */
export async function triggerAutoBackup(): Promise<BackupMetadata | null> {
  try {
    return await invoke<BackupMetadata | null>("trigger_auto_backup");
  } catch (error) {
    console.error("Auto-backup failed:", error);
    return null;
  }
}

/**
 * Clean up old backups based on retention policy
 */
export async function cleanupOldBackups(
  backupDir: string,
  maxBackups: number
): Promise<number> {
  try {
    return await invoke<number>("cleanup_old_backups", { backupDir, maxBackups });
  } catch (error) {
    console.error("Failed to cleanup old backups:", error);
    return 0;
  }
}

/**
 * Format bytes to human-readable string
 */
export function formatBytes(bytes: number): string {
  if (bytes === 0) return "0 B";
  const k = 1024;
  const sizes = ["B", "KB", "MB", "GB"];
  const i = Math.floor(Math.log(bytes) / Math.log(k));
  return `${parseFloat((bytes / Math.pow(k, i)).toFixed(2))} ${sizes[i]}`;
}

/**
 * Format date to human-readable string
 */
export function formatBackupDate(dateString: string): string {
  const date = new Date(dateString);
  const now = new Date();
  const diffMs = now.getTime() - date.getTime();
  const diffDays = Math.floor(diffMs / (1000 * 60 * 60 * 24));

  if (diffDays === 0) {
    const diffHours = Math.floor(diffMs / (1000 * 60 * 60));
    if (diffHours === 0) {
      const diffMins = Math.floor(diffMs / (1000 * 60));
      if (diffMins === 0) return "Just now";
      return `${diffMins} minute${diffMins === 1 ? "" : "s"} ago`;
    }
    return `${diffHours} hour${diffHours === 1 ? "" : "s"} ago`;
  } else if (diffDays === 1) {
    return "Yesterday";
  } else if (diffDays < 7) {
    return `${diffDays} days ago`;
  } else {
    return date.toLocaleDateString(undefined, {
      year: "numeric",
      month: "short",
      day: "numeric",
      hour: "2-digit",
      minute: "2-digit",
    });
  }
}

/**
 * Get the next scheduled backup date based on settings
 */
export function getNextBackupDate(
  lastBackup: Date | null,
  frequency: "daily" | "weekly" | "monthly"
): Date {
  const now = new Date();
  const next = lastBackup ? new Date(lastBackup) : now;

  switch (frequency) {
    case "daily":
      next.setDate(next.getDate() + 1);
      break;
    case "weekly":
      next.setDate(next.getDate() + 7);
      break;
    case "monthly":
      next.setMonth(next.getMonth() + 1);
      break;
  }

  return next > now ? next : now;
}

/**
 * Initialize auto-backup on app start
 * Should be called when the app initializes
 */
export async function initializeAutoBackup(): Promise<void> {
  const settings = loadBackupSettings();

  if (!settings.enabled) {
    return;
  }

  // Set default backup dir if not set
  if (!settings.backup_dir) {
    settings.backup_dir = await getDefaultBackupDir();
    saveBackupSettings(settings);
  }

  // Sync settings with Rust backend
  await setAutoBackupSettings(settings);

  // Check if backup is needed
  await triggerAutoBackup();
}

File: likho-app/src-tauri/src/commands/backup.rs
//! Backup commands for Tauri
//!
//! These commands provide the interface between the frontend and the backup system.

use tauri::State;
use crate::AppState;
use crate::backup::{BackupFormat, BackupService};
use crate::models::{
    AutoBackupSettings, AutoBackupStatus, BackupInfo, BackupMetadata, 
    ExportBackupRequest, ImportBackupRequest, RestorePreview
};
use std::path::PathBuf;
use std::sync::Mutex;

/// Get the current database path
#[tauri::command]
pub fn get_database_path(state: State<AppState>) -> Result<String, String> {
    Ok(state.db_path.to_string_lossy().to_string())
}

/// Export database to a backup file
#[tauri::command]
pub fn export_database(
    state: State<AppState>,
    request: ExportBackupRequest,
) -> Result<BackupMetadata, String> {
    let service = BackupService::new(state.db_path.clone());
    
    let format = match request.format.as_str() {
        "json" => BackupFormat::Json,
        "gzip" => BackupFormat::Gzip,
        _ => BackupFormat::Sqlite,
    };
    
    let target_path = PathBuf::from(&request.target_path);
    
    service.export_database(&target_path, format, request.include_embeddings)
        .map_err(|e| format!("Export failed: {}", e))
}

/// Import database from a backup file (preview only)
#[tauri::command]
pub fn preview_restore(
    state: State<AppState>,
    request: ImportBackupRequest,
) -> Result<RestorePreview, String> {
    let service = BackupService::new(state.db_path.clone());
    let source_path = PathBuf::from(&request.source_path);
    
    service.import_database(&source_path)
        .map_err(|e| format!("Failed to preview restore: {}", e))
}

/// Execute database restore
#[tauri::command]
pub fn execute_restore(
    state: State<AppState>,
    request: ImportBackupRequest,
) -> Result<(), String> {
    let service = BackupService::new(state.db_path.clone());
    let source_path = PathBuf::from(&request.source_path);
    
    service.execute_restore(&source_path)
        .map_err(|e| format!("Restore failed: {}", e))
}

/// Verify backup file integrity
#[tauri::command]
pub fn verify_backup_integrity(
    state: State<AppState>,
    path: String,
) -> Result<bool, String> {
    let service = BackupService::new(state.db_path.clone());
    let backup_path = PathBuf::from(&path);
    
    service.verify_backup_integrity(&backup_path)
        .map_err(|e| format!("Verification failed: {}", e))
}

/// List all backups in a directory
#[tauri::command]
pub fn list_backups(
    state: State<AppState>,
    backup_dir: String,
) -> Result<Vec<BackupInfo>, String> {
    let service = BackupService::new(state.db_path.clone());
    let dir = PathBuf::from(&backup_dir);
    
    service.list_backups(&dir)
        .map_err(|e| format!("Failed to list backups: {}", e))
}

/// Delete a backup file
#[tauri::command]
pub fn delete_backup(
    state: State<AppState>,
    path: String,
) -> Result<(), String> {
    let service = BackupService::new(state.db_path.clone());
    let backup_path = PathBuf::from(&path);
    
    service.delete_backup(&backup_path)
        .map_err(|e| format!("Failed to delete backup: {}", e))
}

// Auto-backup state management
pub struct AutoBackupState {
    settings: Mutex<AutoBackupSettings>,
}

impl AutoBackupState {
    pub fn new() -> Self {
        Self {
            settings: Mutex::new(AutoBackupSettings {
                enabled: false,
                frequency: "weekly".to_string(),
                backup_dir: String::new(),
                max_backups: 10,
                include_embeddings: true,
                format: "sqlite".to_string(),
            }),
        }
    }
}

/// Get auto-backup settings
#[tauri::command]
pub fn get_auto_backup_settings(
    state: State<AutoBackupState>,
) -> Result<AutoBackupSettings, String> {
    let settings = state.settings.lock().unwrap();
    Ok(settings.clone())
}

/// Update auto-backup settings
#[tauri::command]
pub fn set_auto_backup_settings(
    state: State<AutoBackupState>,
    settings: AutoBackupSettings,
) -> Result<(), String> {
    let mut current = state.settings.lock().unwrap();
    *current = settings;
    Ok(())
}

/// Get auto-backup status
#[tauri::command]
pub fn get_auto_backup_status(
    state: State<AppState>,
    settings_state: State<AutoBackupState>,
) -> Result<AutoBackupStatus, String> {
    let settings = settings_state.settings.lock().unwrap();
    let service = BackupService::new(state.db_path.clone());
    
    let backups = if !settings.backup_dir.is_empty() {
        service.list_backups(&PathBuf::from(&settings.backup_dir))
            .unwrap_or_default()
    } else {
        Vec::new()
    };
    
    let last_backup = backups.first().map(|b| b.created_at);
    let backups_count = backups.len() as i32;
    let total_size = backups.iter().map(|b| b.size).sum();
    
    // Calculate next backup time based on frequency
    let next_backup = last_backup.map(|last| {
        use chrono::Duration;
        match settings.frequency.as_str() {
            "daily" => last + Duration::days(1),
            "weekly" => last + Duration::weeks(1),
            "monthly" => last + Duration::days(30),
            _ => last + Duration::weeks(1),
        }
    });
    
    Ok(AutoBackupStatus {
        enabled: settings.enabled,
        last_backup,
        next_backup,
        backups_count,
        total_size,
    })
}

/// Trigger an auto-backup if needed
#[tauri::command]
pub async fn trigger_auto_backup(
    state: State<'_, AppState>,
    settings_state: State<'_, AutoBackupState>,
) -> Result<Option<BackupMetadata>, String> {
    let settings = settings_state.settings.lock().unwrap().clone();
    
    if !settings.enabled || settings.backup_dir.is_empty() {
        return Ok(None);
    }
    
    let service = BackupService::new(state.db_path.clone());
    
    // Check if backup is needed
    let backups = service.list_backups(&PathBuf::from(&settings.backup_dir))
        .unwrap_or_default();
    
    let should_backup = if let Some(last) = backups.first() {
        use chrono::{Duration, Utc};
        let elapsed = Utc::now() - last.created_at;
        match settings.frequency.as_str() {
            "daily" => elapsed >= Duration::days(1),
            "weekly" => elapsed >= Duration::weeks(1),
            "monthly" => elapsed >= Duration::days(30),
            _ => elapsed >= Duration::weeks(1),
        }
    } else {
        true // No backups yet
    };
    
    if !should_backup {
        return Ok(None);
    }
    
    // Generate backup filename
    let timestamp = chrono::Local::now().format("%Y%m%d_%H%M%S");
    let extension = match settings.format.as_str() {
        "json" => "json",
        "gzip" => "db.gz",
        _ => "db",
    };
    let filename = format!("likho_backup_{}.{}_in_progress", timestamp, extension);
    let target_path = PathBuf::from(&settings.backup_dir).join(&filename);
    
    // Perform backup
    let format = match settings.format.as_str() {
        "json" => BackupFormat::Json,
        "gzip" => BackupFormat::Gzip,
        _ => BackupFormat::Sqlite,
    };
    
    let metadata = service.export_database(&target_path, format, settings.include_embeddings)
        .map_err(|e| format!("Auto-backup failed: {}", e))?;
    
    // Rename to remove "_in_progress" suffix
    let final_filename = format!("likho_backup_{}.{}", timestamp, extension);
    let final_path = PathBuf::from(&settings.backup_dir).join(&final_filename);
    let _ = std::fs::rename(&target_path, &final_path);
    
    // Rename metadata file too
    let meta_in_progress = target_path.with_extension("meta.json");
    let meta_final = final_path.with_extension("meta.json");
    let _ = std::fs::rename(&meta_in_progress, &meta_final);
    
    // Clean up old backups
    if backups.len() >= settings.max_backups as usize {
        for old_backup in &backups[settings.max_backups as usize - 1..] {
            let _ = service.delete_backup(&PathBuf::from(&old_backup.path));
        }
    }
    
    Ok(Some(metadata))
}

/// Clean up old backups based on retention policy
#[tauri::command]
pub fn cleanup_old_backups(
    state: State<AppState>,
    backup_dir: String,
    max_backups: i32,
) -> Result<i32, String> {
    let service = BackupService::new(state.db_path.clone());
    let dir = PathBuf::from(&backup_dir);
    
    let backups = service.list_backups(&dir)
        .map_err(|e| format!("Failed to list backups: {}", e))?;
    
    let mut deleted = 0;
    if backups.len() > max_backups as usize {
        for old_backup in &backups[max_backups as usize..] {
            if service.delete_backup(&PathBuf::from(&old_backup.path)).is_ok() {
                deleted += 1;
            }
        }
    }
    
    Ok(deleted)
}
